### DISCUSSION TITLE
Problem rendering Japanese/Chinese characters in status text.

### DISCUSSION THREAD

**Discussion by kodbuzz** [sca075/mqtt_vacuum_camera#87] [Ideas]
### Checklist

- [X] I have updated the integration to the latest version available
- [X] I have checked if the problem is already reported

### The problem

Japanese and Chinese characters are displayed incorrectly when the room is named in these languages.

![image](https://github.com/sca075/valetudo_vacuum_camera/assets/77547202/f3a7d499-a9c2-4f6c-9754-277b9f8aa7f6)

I attempted to resolve the issue using 'Pillow' with the specified Noto-CJK font from 'https://github.com/notofonts/noto-cjk/'. While this approach worked, it resulted in the loss of the charging indicator letter (lightning). Additionally, adding an extra font file, which is approximately 60MB, it does present another challenge imho.

Is there any other approach that could effectively address this issue?

![image](https://github.com/sca075/valetudo_vacuum_camera/assets/77547202/f5ae37cc-2a9e-4b78-afa8-ae3d05e0a069)
This image shows the results using 'Pillow' with font Noto-CJK.

### What version of an integration has described problem?

1.5.8

### What vacuum model do you have problems with?

Mi Robot Vacuum-Mop P (Viomi v7)

### Please firmware installed on your Vacuum.

0047 - Valetudo 2023.12.0

### What type of platform you use?

VEMLinux (Virtual Machine) such as Proxmox, VMWare, etc.

### What version of Home Assistant do you use?

2024.2.4

### What type of installation are you running?

Home Assistant OS

**Comment by sca075** [comment]
@kodbuzz thanks as usual, I will try to fix it, as it is a limitation with the characters (I did change the fonts today) would you please double check downloading the Main please?

**Comment by kodbuzz** [comment]
> @kodbuzz thanks as usual, I will try to fix it, as it is a limitation with the characters (I did change the fonts today) would you please double check downloading the Main please?

@sca075 Good evening, since I'm on the latest commit, which was made approx 5hrs ago, it seems that the issue still persists. Btw I'm not certain if Cyrillic alphabets encounter the same rendering issue, shall we test with Cyrillic as well?

**Comment by sca075** [comment]
@kodbuzz of course if you found a font for it would be appreciated if you could share it so that it will be used. Just as question. Lato was previously working with Japanese?

**Comment by kodbuzz** [comment]
@sca075 unfortunately no, afaik the "status_text" field has never rendered correctly when it contains Japanese/Chinese characters even in previous versions, It seems like more of an aesthetic issue to me, so I didn't submit the problem in the past few weeks as I was quite occupied too.

**Comment by sca075** [comment]
@kodbuzz okay we will found a way to make it work 😉

**Comment by sca075** [comment]
@kodbuzz release 1.5.9-rc1 include a font that can show the Japanese characters
![Screenshot 2024-02-28 at 22 45 05](https://github.com/sca075/valetudo_vacuum_camera/assets/82227818/f978d37b-89f4-4e5f-8fd7-772bc2743401)
If you confirm it I will use this font as soon I release the 1.5.9 ;)

**Comment by kodbuzz** [comment]
@sca075 confirmed, it works like a charm, thank you!

**Comment by sca075** [comment]
> @sca075 confirmed, it works like a charm, thank you!
@kodbuzz I tested it and it looks Hiragana, Katakana is probably okay but Kanji as well Cyrillic on the status text could be a little challenging but not impossible.. for the time be if we are okay with it will use this font later on probably I can try to implement a menu were to select the language (font) to use for the status test.. will close the issue but open a discussion as usual to keep trace of this idea, thanks again.

**Reply by kodbuzz** [reply]
Kanjis are still shown as hollow squares unfortunately, I found that it worked because of the I forgot to change font from NotoSansCJKjp-VF.ttf to FreeSerif.ttf but simply copied FreeSerif.ttf to fonts folder, my bad. Do you mind have a look at 

https://github.com/notofonts/noto-cjk/tree/main/Sans/Variable/TTF

The fonts belong to the Noto family and are licensed under the SIL Open Font License (OFL). However, the downside is their size, which poses a challenge.

**Reply by kodbuzz** [reply]
@sca075 
The font resolves the rendering issues for Japanese, Chinese, and Korean characters simultaneously, but it lacks support for the Unicode symbol u2211...

**Reply by sca075** [reply]
Yep this is one of problems, need to found another symbol for charge status, but I did some research and apparently one possible solution (that I didn't try yet) is to get the list of the fonts installed on the OS and then.. an universal Unicode symbol for the charge status, so that changes on status text with the room name will be displayed with the user font choice.

Or.. we need to write our own font set 😂 and I'm not willing to do it.

Anyhow @kodbuzz as usual you raised a really good point and I need to thank you for that.

Have a wonderful evening.

**Reply by sca075** [reply]
> Yep this is one of problems, need to found another symbol for charge status, but I did some research and apparently one possible solution (that I didn't try yet) is to get the list of the fonts installed on the OS and then.. an universal Unicode symbol for the charge status, so that changes on status text with the room name will be displayed with the user font choice.
> 
> Or.. we need to write our own font set 😂 and I'm not willing to do it.
> 
> Anyhow @kodbuzz as usual you raised a really good point and I need to thank you for that.
> 
> Have a wonderful evening.

P.s. probably an up arrow ⬆️ for changing and low arrow for ⬇️ discharge will be the new battery symbols.

**Comment by sca075** [comment]
@kodbuzz I think we have a possible compromise, will use both fonts it's jus a string operation more to add.

**Reply by kodbuzz** [reply]
@sca075 
I'm considering the possibility of detecting the character sets in the "status_text" and downloading only the necessary fonts corresponding to them when needed. This way, users who are not involved won't be bothered, and the integration itself won't need to include extra fonts to keep in a compact and decent size. but I'm unsure if HACS allows fetching and storing external resources.


    def is_cjk_character(char):
        """
        Checks if a character belongs to specific CJK character ranges.
        """
        return (
            ('\u4e00' <= char <= '\u9fff')  # Chinese
            or ('\u3040' <= char <= '\u309f')  # Hiragana
            or ('\u30a0' <= char <= '\u30ff')  # Katakana
            or ('\uac00' <= char <= '\ud7af')  # Korean Hangul
        )
        
        
Maybe you would like to incorporate additional detections here to allow for future extensibility, especially if more languages are supported down the line in the future?

**Reply by sca075** [reply]
I would instead use the font you provided with weight =1 and we use your character only inside (rooms) 
This way we can still use the original font, and we don't need to change the text icons;)

**Reply by sca075** [reply]
And yes I will include more languages in the future 😊

**Reply by sca075** [reply]
@kodbuzz please download the main and let me know when it works also for you ;) The modification is done.

**Reply by kodbuzz** [reply]
@sca075 thank you so much! I've just downloaded the latest version from the main branch, and oddly, I noticed that the room name isn't displayed at all as expected. I haven't figured out what's causing this yet. I'll check if it's an issue on my end or something else once I'm home.

**Reply by kodbuzz** [reply]
Got

`  in_room: null`

I don't have any clue how this happened, the map appears to be completely fine and so are the positions of both the robot and the docker. I will try to figure it out ~~tomorrow morning, about 12 hrs later.~~ Tested, all works!

**Reply by kodbuzz** [reply]
@sca075 Hey! Just wanted to let you know that you nailed it! 🎉🚀

![image](https://github.com/sca075/valetudo_vacuum_camera/assets/77547202/6225a951-59a8-4703-98e2-3329db7c2bc9)
